libraries/chain/apply_context.cpp

```diff
@@ -29,68 +30,98 @@ static inline void print_debug(account_name receiver, const action_trace& ar) {
    }
 }
 
-void apply_context::exec_one( action_trace& trace )
+apply_context::apply_context(controller& con, transaction_context& trx_ctx, uint32_t action_ordinal, uint32_t depth)
+:control(con)
+,db(con.mutable_db())
+,trx_context(trx_ctx)
+,recurse_depth(depth)
+,first_receiver_action_ordinal(action_ordinal)
+,action_ordinal(action_ordinal)
+,idx64(*this)
+,idx128(*this)
+,idx256(*this)
+,idx_double(*this)
+,idx_long_double(*this)
+{
+   action_trace& trace = trx_ctx.get_action_trace(action_ordinal);
+   act = &trace.act;
+   receiver = trace.receiver;
+   context_free = trace.context_free;
+}
+
+void apply_context::exec_one()
 {
    auto start = fc::time_point::now();
 
    action_receipt r;
    r.receiver         = receiver;
-   r.act_digest       = digest_type::hash(act);
-
-   trace.trx_id = trx_context.id;
-   trace.block_num = control.pending_block_state()->block_num;
-   trace.block_time = control.pending_block_time();
-   trace.producer_block_id = control.pending_producer_block_id();
-   trace.act = act;
-   trace.context_free = context_free;
+   r.act_digest       = digest_type::hash(*act);
 
    const auto& cfg = control.get_global_properties().configuration;
+   const account_metadata_object* receiver_account = nullptr;
    try {
       try {
-         const auto& a = control.get_account( receiver );
-         privileged = a.privileged;
-         auto native = control.find_apply_handler( receiver, act.account, act.name );
+         receiver_account = &db.get<account_metadata_object,by_name>( receiver );
+         privileged = receiver_account->is_privileged();
+         auto native = control.find_apply_handler( receiver, act->account, act->name );
          if( native ) {
             if( trx_context.enforce_whiteblacklist && control.is_producing_block() ) {
                control.check_contract_list( receiver );
-               control.check_action_list( act.account, act.name );
+               control.check_action_list( act->account, act->name );
             }
             (*native)( *this );
          }
 
-         if( a.code.size() > 0
-             && !(act.account == config::system_account_name && act.name == N( setcode ) &&
-                  receiver == config::system_account_name) ) {
+         if( ( receiver_account->code_hash != digest_type() ) &&
+               ( control.is_builtin_activated( builtin_protocol_feature_t::forward_setcode )
+                  || !( act->account == config::system_account_name
+                        && act->name == N( setcode )
+                        && receiver == config::system_account_name )
+               )
```

Comment:
Seems like the right-hand side of `||` would be faster to check. Can we check it first.

Reply:
Precedence order is `&&` then `||` but I think another set of parens would make it clearer.

Reply:
Resolved by https://github.com/EOSIO/eos/pull/7147/commits/a0f40f517b150f9a3162068aa032c3a95f721332.